File: mindspore/python/mindspore/ops/_op_impl/_custom_op/flash_attention/tiling_strategy/strategy.py
# Copyright 2023 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""The base class of tiling strategy"""
from abc import ABCMeta
from abc import abstractmethod
from collections import namedtuple

TilingPara = namedtuple("TilingPara", "Br last_Br Bc last_Bc Tr Tc")


class TilingStrategy(metaclass=ABCMeta):
    """Tiling strategy interface. All implementations should be defined in this module,
    otherwise, the UT will fail.
    """

    _strategies = {}

    def __init__(self, Nq, N, head_dim) -> None:
        super().__init__()
        self.Nq = Nq
        self.N = N
        self.Br = None
        self.last_Br = None
        self.Bc = None
        self.last_Bc = None
        self.Tr = None
        self.Tc = None
        self.d = head_dim

    def __init_subclass__(cls, **kwargs):
        TilingStrategy._strategies[cls.strategy_name()] = cls

    @classmethod
    @abstractmethod
    def strategy_name(cls):
        """strategy name"""
        raise NotImplementedError

    @classmethod
    def from_strategy_name(cls, stgy_name: str):
        """from strategy name"""
        stgy_clz = TilingStrategy._strategies.get(stgy_name)
        if stgy_clz is None:
            raise Exception(f"Strategy:{stgy_name} not supported")

        return stgy_clz

    @abstractmethod
    def tiling(self) -> TilingPara:
        """tiling"""
        raise NotImplementedError

    def gen_tiling_para(self) -> TilingPara:
        """gen tiling para"""
        return TilingPara(self.Br, self.last_Br, self.Bc, self.last_Bc, self.Tr, self.Tc)

File: mindspore/ccsrc/pipeline/jit/pi_jit/graph_guard/strategy.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "pipeline/jit/pi_jit/graph_guard/strategy.h"

namespace mindspore {
namespace jit {
namespace graph {

OptStrategy::ExecKind OptStrategy::MakeExecStrategyByPerf(OptPerfPtr graph_perf, OptPerfPtr pynative_perf,
                                                          double adj_coef) {
  PerfStatisticsPtr graph_stat = graph_perf->GetStatistics();
  PerfStatisticsPtr pynative_stat = graph_perf->GetStatistics();
  if (graph_stat->GetAverageDuration() * (1 + adj_coef) > pynative_stat->GetAverageDuration()) {
    return ExecKind::kExecPyNative;
  } else {
    return ExecKind::kExecGraph;
  }
}

OptStrategy::ExecKind OptStrategy::MakeExecStrategyByComplex(PyObject *code, int threshold) {
  PyCodeObject *co = nullptr;
  if (PyCode_Check(code)) {
    co = reinterpret_cast<PyCodeObject *>(code);
  } else if (PyFunction_Check(code)) {
    co = reinterpret_cast<PyCodeObject *>(PyFunction_GET_CODE(code));
  }
  // currently just use instruction count to judge whether to use graph build
  // later it need cost model to make judgement here
  if (co != nullptr && static_cast<int>(PyBytes_GET_SIZE(co->co_code) / sizeof(_Py_CODEUNIT)) < threshold) {
    return ExecKind::kExecPyNative;
  } else {
    return ExecKind::kExecGraph;
  }
}
}  // namespace graph
}  // namespace jit
}  // namespace mindspore
